datasets/flwr_datasets/partitioner/dirichlet_partitioner.py

```diff
@@ -0,0 +1,300 @@
+                # than the average numbers of samples per partition. Note that it might
+                # especially affect classes that are later in the order. This is the
+                # reason for more sparse division that the alpha might suggest.
+                if self._self_balancing:
+                    assert self._avg_num_of_samples_per_node is not None
+                    for nid in nid_to_proportion_of_k_samples.copy():
+                        if (
+                            len(node_id_to_indices[nid])
+                            > self._avg_num_of_samples_per_node
+                        ):
+                            nid_to_proportion_of_k_samples[nid] = 0
+
+                    # Normalize the proportions such that they sum up to 1
+                    sum_proportions = sum(nid_to_proportion_of_k_samples.values())
+                    for nid, prop in nid_to_proportion_of_k_samples.copy().items():
+                        nid_to_proportion_of_k_samples[nid] = prop / sum_proportions
+
+                # Determine the split indices
+                cumsum_division_fractions = np.cumsum(
+                    list(nid_to_proportion_of_k_samples.values())
+                )
+                cumsum_division_numbers = cumsum_division_fractions * len(
+                    indices_representing_class_k
+                )
+                # [:-1] is because the np.split requires the division indices but the
+                # last element represents the sum = total number of samples
+                indices_on_which_split = cumsum_division_numbers.astype(int)[:-1]
+
+                split_indices = np.split(
+                    indices_representing_class_k, indices_on_which_split
+                )
+
+                # Append new indices (coming from class k) to the existing indices
+                for nid, indices in node_id_to_indices.items():
+                    indices.extend(split_indices[nid].tolist())
+
+            # Determine if the indices assignment meets the min_partition_size
+            # If it does not mean the requirement repeat the Dirichlet sampling process
+            # Otherwise break the while loop
+            min_sample_size_on_client = min(
+                len(indices) for indices in node_id_to_indices.values()
+            )
+            if min_sample_size_on_client >= self._min_partition_size:
+                break
+            warnings.warn(
+                f"The specified min_partition_size of the create the "
+                f"partitions was not satisfied as the direct result of the "
+                f"{sampling_try} st/nd/rd/th sampling from the Dirichlet "
+                f"distribution. The probability sampling from the Dirichlet "
+                f"distribution will be repeated. Note: It is not a desired "
+                f"behavior. It is recommended to adjust the alpha or "
+                f"min_partition_size instead.",
+                stacklevel=1,
+            )
```

Comment:
```suggestion
            sample_sizes = [len(indices) for indices in node_id_to_indices.values()]
            alpha_not_met = [
                self._alpha[i]
                for i, ss in enumerate(sample_sizes)
                if ss == min(sample_sizes)
            ]
            mssg_list_alphas = (
                (
                    " Generating partitions by sampling from a list of very"
                    "dispair alpha values can be hard to achieve. "
                    f"Try reducing the range between maximum ({max(self._alpha)}) and "
                    f"minimum alpha ({min(self._alpha)}) values."
                )
                if len(self._alpha.flatten().tolist()) > 0
                else ""
            )
            warnings.warn(
                f"The specified min_partition_size ({self._min_partition_size}) was "
                f"not satisfied for alpha ({alpha_not_met}) after "
                f"{sampling_try} attempts at sampling from the Dirichlet "
                f"distribution. The probability sampling from the Dirichlet "
                f"distribution will be repeated. Note: This is not a desired "
                f"behavior. It is recommended to adjust the alpha or "
                f"min_partition_size instead. {mssg_list_alphas}",
                stacklevel=1,
            )
```

Reply:
I thought of adding a bit more info to the warning. I think problems will mostly occur when someone tries to pass `alpha=[1000,0.1, 1000,1000]` (which has a wide range between `min` and `max) and then asking for a high value of `min_partitions`.

Reply:
I agree that would be the main cause and solution. Yet it can also be mitigated to some degree if someone specifies the array but multiplies all the numbers by a constant so e..g  * 1_000 `[1_000_000, 100, 1_000_000, 1_000_000]`. But I'm not sure if that should also be added.